packages/block-editor/src/components/use-setting/index.js

```diff
@@ -113,42 +118,69 @@ export default function useSetting( path ) {
 				);
 				return undefined;
 			}
-			const settings = select( blockEditorStore ).getSettings();
 
-			// 1 - Use __experimental features, if available.
-			// We cascade to the all value if the block one is not available.
+			let result;
 			const normalizedPath = removeCustomPrefixes( path );
-			const defaultsPath = `__experimentalFeatures.${ normalizedPath }`;
-			const blockPath = `__experimentalFeatures.blocks.${ blockName }.${ normalizedPath }`;
-			const experimentalFeaturesResult =
-				get( settings, blockPath ) ?? get( settings, defaultsPath );
 
-			if ( experimentalFeaturesResult !== undefined ) {
+			// 1. Take settings from the block instance or its ancestors.
```

Comment:
The algorith for `useSetting` is:

- lookup for settings defined in the block instance or any of its ancestors => this is the only behavioral change introduced by this PR.
- lookup for settings defined in the block editor store
- fallback for deprecated settings
- fallback for dropcap

Reply:
In global styles, I added a custom palette to the "paragraph" block and then selected a paragraph block from your example above. I wasn't sure what should show up: 
 
 - The palette of the parent "group"
 - The palette of the "paragraph" block from Global styles.

I think that's the remaining point we should figure out in the algorithm, we should probably just pick one approach and stick with it. 

My first thought is that picking the "paragraph" settings from the global styles might make more sense here. (Right now it picks the parent's palette)

cc @jorgefilipecosta @mcsf

Reply:
I shared here some longer thoughts at https://github.com/WordPress/gutenberg/issues/40318#issuecomment-1106694527 

I've looked at use cases for sections, aka why do we want to do this? I understand we want to enable doing things such as "set colors for this whole section" or "disable custom font sizes for the whole section". These are easier to do with a "by source" algorithm (first considers block instance and its ancestors, then theme.json data). Specially when/if we expose this to the users via an UI.

Reply:
I think it's not that simple. The current algorithm is different that "style" and it's not clear why we'd want "settings" to work differently.

If you apply a text color to a paragraph block, it won't be using the color from the parent "style", it will continue to use the "global styles" one.

Reply:
Yeah, but styles and settings behave different in `theme.json` as well: styles aren't inherited and settings are .

Reply:
Some "styles" are inherited (CSS inheritance :)) but I see your point. I can be onboard with the current algorithm. I know @jorgefilipecosta had thoughts about this as well, so let's just agree and move forward.

Reply:
I'm on the opinion of @youknowriad, I think first a block would use settings specific for that block, (at the parent level or at the theme.json level) and if there are block specific settings use them if not use the global settings first from the ancestor and them from theme.json
I think the algorithm I referred to follows what happens naturally with CSS imagine we have the red color defined at the global level as red-1, for wp-block-child as red-2 and then there is a block wp-block-parent that is section and defines red as red-3 we would have more or less this output
```
// theme.json
body {
--wp--preset--color--red: red-1;
}

.wp-block-child {
--wp--preset--color--red: red-2;
}

// settings mechanism local to the block
.wp-container-section-block-parent {
--wp--preset--color--red: red-2;
}

//html
<div class="...wp-container-section-block-parent...><div class="my-child-block wp-block-child>What is wp--preset--color--red?</div></div>
```
According to the browser wp--preset--color--red on my-child-block/wp-block-child is going to be red-2. The block-specific color defined at the theme.json. I think we should follow the same model it may make things easier for developers.

Reply:
> Following the CSS model. We need to lock them down for every block type, otherwise, an external source could affect it.

Setting a setting specifically for the block is not something so common most themes will not do it. And in the cases, a theme or a plugin does it would be ok to inside the pattern follow what the theme decided for that specific block because it is a specific situation. For example, imagine on the table block because it has the stripes option I set a color palette for the table block that is stripe friendly like the block had in the past. I think if the table block is used inside a section that has a color palette it would still be ok to use the table block palette because the table block is a specific situation. If the pattern really wants to overwrite the table block palette it would be ok for fore the pattern to set a block-specific palette.
The CSS model allows us to choose between still keeping the table palette or overwriting anyway. On the alternative model, I don't really have a choice. Imagine I really wanted that the table block kept its palette although for all the other blocks in my section I'm setting a different palette, I will not be able to do that. I will not be able to keep theme-specific settings and I think in some cases it may make sense to do that.

But it is not an easy decision there are advantages and disadvantages to both options.

If we don't follow the CSS model what is the solution we are going to have to on the example I gave `<div class="...wp-container-section-block-parent...><div class="my-child-block wp-block-child>What is wp--preset--color--red?</div></div>`make  wp--preset--color--red have the value red-3 instead of red-2?
I think we need to find a solution to this issue if we are going for a solution where the CSS model does not work by default.

Reply:
> The CSS model allows us to choose between still keeping the table palette or overwriting anyway. On the alternative model, I don't really have a choice.

You do! You just set the settings for the specific block types you're interested in (and the other ones -the table- will fall back to the store settings).

I think the crux of the conversation is: what are the common uses case we should optimize for? If [this one](https://github.com/WordPress/gutenberg/pull/40547#discussion_r859537860) is the typical thing patterns want to do and the table example you shared is the uncommon one, the model this PR follows works best.

> what is the solution we are going to have to on the example I gave

I've created a pen that demonstrates the challenge you shared and how we can address it https://codepen.io/oandregal/pen/eYyqxZO?editors=1100

Let me unwrap it here as well. Given a section like this one (note the paragraph has a red color applied by the user):

```html
  <div class="wp-section-UUID">
    <p class="has-red-color"></p>
  </div>
```

If the theme has added some paragraph-specific settings, as in:

```json
{
  "settings": {
    "color": {
      "palette": [
	    { "slug": "red", "color": "red-theme-global" }
      ]
    },
    "blocks": {
      "core/paragraph": {
        "color": {
          "palette": [
            { "slug": "red", "color": "red-theme-block" }
          ]
        }
      }
    }
  }
}
```

And the section has this:

```json
{
  "settings": {
    "color": {
      "palette": [
        { "slug": "red", "color": "red-section-global" }
      ]
    }
  }
}
```

**How do we make it so the color rendered for the paragraph is `red-section-global`?**

```css
/*
 * THIS IS THE CSS GENERATED FOR THE `theme.json`
 */
body { --wp--preset--color--red: "red-theme-global"; }
p { --wp--preset--color--red: "red-theme-block"; }
.has-red-color { color: var(--wp--preset--color--red) !important; }

/*
 * THIS IS THE CSS GENERATED FOR THE SECTION.
 */
.wp-section-UUID,
.wp-section-UUID [class*="wp-block"], /* Any child block */
/*
 * We also need to cover for child blocks that don't use the .wp-block-* class:
 * paragraph, list (ul, or), and heading (h1-h6).
 */
.wp-section-UUID p,
.wp-section-UUID li,
.wp-section-UUID ul,
.wp-section-UUID h1,
.wp-section-UUID h2,
.wp-section-UUID h3,
.wp-section-UUID h4,
.wp-section-UUID h5,
.wp-section-UUID h6,
{
  --wp--preset--color--red: hotpink;
}
```

Does this address your concerns?

Note that this PR doesn't generate the styles for the presets declared in a section. That's something that has a complexity of its own and I thought that to be a follow-up (things like generating the preset classes or not depending on whether the preset is already defined in `theme.json`, etc).

Reply:
Ongoing draft PR for rendering the settings at https://github.com/WordPress/gutenberg/pull/40683

Reply:
> You do! You just set the settings for the specific block types you're interested in (and the other ones -the table- will fall back to the store settings).


Your section may not be aware of the blocks it is interested in, it may have a group where the user can nest anything inside.

> I think the crux of the conversation is: what are the common uses case we should optimize for? If https://github.com/WordPress/gutenberg/pull/40547#discussion_r859537860 is the typical thing patterns want to do and the table example you shared is the uncommon one, the model this PR follows works best.

I agree we should optimize for the most common cause. The thing is we don't know the most common case yet. I think a theme setting block specific settings is something rare and very specific that should be respected even if that block is used inside a section setting something global for that setting. Your expectation is that something global at the section should take priority above a theme-specific setting. I guess we need to choose and iterated and if we receive feedback or new information later change the direction if that's a possibility.


> How do we make it so the color rendered for the paragraph is red-section-global?
```
.wp-section-UUID,
.wp-section-UUID [class*="wp-block"], /* Any child block */
/*
 * We also need to cover for child blocks that don't use the .wp-block-* class:
 * paragraph, list (ul, or), and heading (h1-h6).
 */
.wp-section-UUID p,
.wp-section-UUID li,
.wp-section-UUID ul,
.wp-section-UUID h1,
.wp-section-UUID h2,
.wp-section-UUID h3,
.wp-section-UUID h4,
.wp-section-UUID h5,
.wp-section-UUID h6,
```

The solution proposed makes things work but requires us to ship much more CSS code, for every color or font size, etc instead of just need to use just .wp-section-UUID, we need to use a much bigger selector per preset we have that means more bytes sent over the wire for every section that has some preset.


 I'm not sure of the best approach either so I'm fine with giving a try to the solution we have here and seeing how it works in practice.

Reply:
Actually, there's a simpler CSS for the presets generated for the section if we embrace how `theme.json` works:

```css
/*
 * CSS Custom Properties defined in the section
 */
.wp-section-UUID .has-red-color {
  --wp--preset--color--red: hotpink;
}

/*
 * CSS classes for presets defined in the section.
 * Ideally, we should not enqueue the classes that are already enqueued by theme.json.
 * Though we still need to check at runtime because a given theme may have them
 * but a different theme may not.
*/
.has-red-color { color: var(--wp--preset--color--red) !important; } 
```

That's it :)

Reply:
Yes but if have a block specific color, on theme.json we generate: 
p.has-red-color {
color: var(--wp--preset--color--red
}

So in each section we can not do just:
```
.wp-section-UUID .has-red-color {
  --wp--preset--color--red: "red-section-global";
}
```
We also need to overwrite the block specific colors:

```
.wp-section-UUID p.has-red-color {
  --wp--preset--color--red: "red-section-global";
}
```